staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -189,9 +196,16 @@ type Backoff struct {
 //
 // If the condition never returns true, ErrWaitTimeout is returned. All other
 // errors terminate immediately.
-func ExponentialBackoff(backoff Backoff, condition ConditionFunc) error {
+//
+// Close stopCh to stop. Pass NeverStop to if you don't want it stop.
+func ExponentialBackoffUntil(backoff Backoff, condition ConditionFunc, stopCh <-chan struct{}) error {
 	duration := backoff.Duration
 	for i := 0; i < backoff.Steps; i++ {
+		select {
+		case <-stopCh:
```

Comment:
shouldn't the stopCh also cancel the sleep below?

Reply:
@sttts That is what I had originally thought. However, here is a convo that I had on #sig-api-machinery a while back:

*begin slack convo*

I have a question concerning the way the `stopCh` is handled throughout the codebase. Consider this example:
```go
go wait.Until(func() {
  glog.Infof("starting metrics server on %v", s.MetricsServer.Addr)
  err := s.MetricsServer.ListenAndServe()
  if err != nil {
    utilruntime.HandleError(fmt.Errorf("failed to start metrics server: %v", err))
  }
}, 5*time.Second, stop)
```
Now, `Until` is a wrapper for `JitterUntil` which looks like this (simplified):
```go
for {
    select {
    case <-stopCh:
        return
    default:
    }

    func() {
        f()
    }()
}
```
The issue is that if `f()`is executed, and the `stopCh` is closed during it’s execution, there is no deterministic way to *guarantee* that the go routine gracefully shutdown and deallocated. In this specific example, `ListenAndServe` will return an error if the listener is closed or an error is encountered. If the `stopCh` is closed, nothing tells either of those two things to happen and so the server will not be closed and if this `stopCh` is in response to a signal handler, than there could be connections to the server that aren’t terminated properly. There are numerous other examples (including one that I’m currently working on in a PR).

*end slack convo*

@liggitt had said:
> For the things that loop infinitely, stopCh just tells them when to stop retrying

In that context, this function is doing just that. The `time.Sleep` will occur even if the `stopCh` is closed. However, when the next iteration of the loop occurs, it will not retry as the `stopCh` has been closed.

Reply:
@sttts does that answer you question about `time.Sleep`?

Reply:
In general we don't pass down the stopCh very deep into the call stack in many cases because, mainly because we don't have contexts on client calls anyway. But we are changing it https://github.com/kubernetes/kubernetes/pull/58677. IMO a sleep is a no-brainer to wire in a channel/context.